lwesp/src/system/lwesp_sys_threadx.c

```diff
@@ -0,0 +1,235 @@
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the follwing conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE
+ * AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * Author:          Tilen MAJERLE <tilen@majerle.eu>
+ * Author:          Neo Xiong <xiongyu0523@gmail.com>
+ * Version:         v1.1.0-dev
+ */
+#include "system/lwesp_sys.h"
+#include "lwesp/lwesp_mem.h"
+#include "tx_api.h"
+
+#if !__DOXYGEN__
+
+#if !defined(LWESP_MEM_SIZE)
+#define LWESP_MEM_SIZE                    0x1000
+#endif
+
+#define LWESP_TICK_PERIOD                 (1000 / TX_TIMER_TICKS_PER_SECOND)
+#define LWESP_MS_TO_TICKS(ms)             ((ms) * TX_TIMER_TICKS_PER_SECOND / 1000 )
+
+static UCHAR        byte_pool_mem[LWESP_MEM_SIZE];
+TX_BYTE_POOL        lwesp_byte_tool;
+static TX_MUTEX     _sys_mutex;
+
+uint8_t
+lwesp_sys_init(void) {
+
+    UINT status;
+    
+    status = tx_byte_pool_create(&lwesp_byte_tool, "byte pool", byte_pool_mem, LWESP_MEM_SIZE);
+    if (status == TX_SUCCESS) {
+        status = tx_mutex_create(&_sys_mutex, "sys mutex", TX_INHERIT);
+    }
+
+    return status == TX_SUCCESS ? 1 : 0;
+}
+
+uint32_t
+lwesp_sys_now(void) {
+    return tx_time_get() * LWESP_TICK_PERIOD;
+}
+
+uint8_t
+lwesp_sys_protect(void) {
+    return tx_mutex_get(&_sys_mutex, TX_WAIT_FOREVER) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_unprotect(void) {
+    return tx_mutex_put(&_sys_mutex) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_create(lwesp_sys_mutex_t* p) {
+    return tx_mutex_create(p, TX_NULL, TX_INHERIT) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_delete(lwesp_sys_mutex_t* p) {
+    return tx_mutex_delete(p) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_lock(lwesp_sys_mutex_t* p) {
+    return tx_mutex_get(p, TX_WAIT_FOREVER) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_unlock(lwesp_sys_mutex_t* p) {
+    return tx_mutex_put(p) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_isvalid(lwesp_sys_mutex_t* p) {
+    return p->tx_mutex_id != TX_CLEAR_ID ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mutex_invalid(lwesp_sys_mutex_t* p) {
+    /* No need actions since all invalid are following delete, and delete make sure it is invalid */
+    return 1;
+}
+
+uint8_t
+lwesp_sys_sem_create(lwesp_sys_sem_t* p, uint8_t cnt) {
+    return tx_semaphore_create(p, TX_NULL, cnt) == TX_SUCCESS ? 1 : 0;
```

Comment:
Please check if this generates correct output. Semaphore must always have only one token, `cnt` variable indicates if this token is free from beginning (`1`) or taken immediately (`0`)

Reply:
I think it is correct. There is no binary semaphore in threadx. The third parameter of `tx_semaphore_create` is the initial count, since you will pass 1 or 0 only, then the semaphore is created in available (1) or (0) unavailable state.